clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,255 @@
+            }
+            return;
+        }
+
+        onError(response, currentTimeMs);
+    }
+
+    private void onError(final ConsumerGroupHeartbeatResponse response,
+                         final long currentTimeMs) {
+
+        this.heartbeatRequestState.onFailedAttempt(currentTimeMs);
+        short errorCode = response.data().errorCode();
+        if (errorCode == Errors.NOT_COORDINATOR.code() || errorCode == Errors.COORDINATOR_NOT_AVAILABLE.code()) {
+            logger.info("GroupHeartbeatRequest failed: coordinator is either not started or not valid. Retrying in " +
+                    "{}ms: {}",
+                heartbeatRequestState.remainingBackoffMs(currentTimeMs),
+                response.data().errorMessage());
+            coordinatorRequestManager.markCoordinatorUnknown(response.data().errorMessage(), currentTimeMs);
+        } else if (errorCode == Errors.COORDINATOR_LOAD_IN_PROGRESS.code()) {
+            // retry
+            logger.info("GroupHeartbeatRequest failed: Coordinator {} is loading. Retrying in {}ms: {}",
+                coordinatorRequestManager.coordinator(),
+                heartbeatRequestState.remainingBackoffMs(currentTimeMs),
+                response.data().errorMessage());
+        } else if (errorCode == Errors.GROUP_AUTHORIZATION_FAILED.code()) {
+            GroupAuthorizationException error = GroupAuthorizationException.forGroupId(membershipManager.groupId());
+            logger.error("GroupHeartbeatRequest failed due to group authorization failure: {}", error.getMessage());
+            nonRetriableErrorHandler.handle(error);
+        } else if (errorCode == Errors.INVALID_REQUEST.code()) {
+            logger.error("GroupHeartbeatRequest failed due to fatal error: {}", response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.INVALID_REQUEST.exception());
+        } else if (errorCode == Errors.GROUP_MAX_SIZE_REACHED.code()) {
+            logger.error("GroupHeartbeatRequest failed due to the max group size limit: {}",
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.GROUP_MAX_SIZE_REACHED.exception());
+        } else if (errorCode == Errors.UNSUPPORTED_ASSIGNOR.code()) {
+            logger.error("GroupHeartbeatRequest failed due to unsupported assignor {}: {}",
+                membershipManager.assignorSelection(), response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNSUPPORTED_ASSIGNOR.exception());
+        } else if (errorCode == Errors.UNRELEASED_INSTANCE_ID.code()) {
+            logger.error("GroupHeartbeatRequest failed due to the instance id {} was not released: {}",
+                membershipManager.groupInstanceId().orElse("null"),
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNRELEASED_INSTANCE_ID.exception());
+        }
+        membershipManager.updateState(response.data());
```

Comment:
When we transition to Failed in updateState, I think that it analogous to a non retriable error. Is our plan to capture all the non-retriable errors before we reach this? We also do some error handling in updateState. The responsibilities are not clear here.

Reply:
Thanks, I made some update to the manager, in particular, I split the fatal error out of the updateState, LMK if you like the change, or we could do better.